src/ur5e_arm.cpp

```diff
@@ -403,6 +403,63 @@ void UR5eArm::send_trajectory(const std::vector<vector6d_t>& p_p,
 
 // helper function to read a data packet and send a noop message
 bool UR5eArm::read_joint_keep_alive() {
+    // check to see if an estop has occurred.
+    std::string status;
+    dashboard->commandSafetyStatus(status);
+
+    if (status.find(urcl::safetyStatusString(urcl::SafetyStatus::NORMAL)) == std::string::npos) {
+        // the arm is currently estopped. save this state.
+        estop.store(true);
+    } else {
+        // the arm is in a normal state.
+        if (estop.load()) {
+            // if the arm was previously estopped, attempt to recover from the estop.
+            // We should not enter this code without the user interacting with the arm in some way(i.e. resetting the estop)
+            try {
+                BOOST_LOG_TRIVIAL(info) << "recovering from e-stop" << std::endl;
+                driver->resetRTDEClient(path_offset + OUTPUT_RECIPE, path_offset + INPUT_RECIPE);
+
+                if (!dashboard->commandPowerOff()) {
+                    BOOST_LOG_TRIVIAL(error)
+                        << "read_joint_keep_alive dashboard->commandPowerOff() returned false when attempting to restart the arm";
+                    return false;
+                }
+                if (!dashboard->commandPowerOn()) {
+                    BOOST_LOG_TRIVIAL(error)
+                        << "read_joint_keep_alive dashboard->commandPowerOn() returned false when attempting to restart the arm";
+                    return false;
+                }
+                // Release the brakes
+                if (!dashboard->commandBrakeRelease()) {
+                    BOOST_LOG_TRIVIAL(error)
+                        << "read_joint_keep_alive dashboard->commandBrakeRelease() returned false when attempting to restart the arm";
+                    return false;
+                }
+                BOOST_LOG_TRIVIAL(info) << "arm restarted, sending control program again" << std::endl;
+
+                // send control script
+                if (!driver->sendRobotProgram()) {
+                    BOOST_LOG_TRIVIAL(error)
+                        << "read_joint_keep_alive driver->sendRobotProgram() returned false when attempting to restart the arm";
+                    return false;
+                }
+                BOOST_LOG_TRIVIAL(info) << "send robot program successful, restarting communication" << std::endl;
+
+            } catch (...) {
+                BOOST_LOG_TRIVIAL(info) << "failed to restart the arm" << std::endl;
+                return false;
+            }
```

Comment:
kept the try catch because `resetRTDEClient` can throw, but I did not confirm that the other functions in the Try do not. I also like the visual that `all of these steps must pass for us to successfully recover` that the try/catch shows